network/src/protocols/discovery/mod.rs

```diff
@@ -1,426 +1,364 @@
+                                    .is_disconnect()
+                                {
+                                    if context.disconnect(session.id).is_err() {
+                                        debug!("disconnect {:?} send fail", session.id)
+                                    }
+                                    return;
+                                }
+                            }
+
+                            if nodes.items.len() > MAX_ADDR_TO_SEND {
+                                warn!(
+                                    "Too many items (announce=false) length={}",
+                                    nodes.items.len()
+                                );
+                                let misbehavior = Misbehavior::TooManyItems {
+                                    announce: nodes.announce,
+                                    length: nodes.items.len(),
+                                };
+
+                                if self
+                                    .addr_mgr
+                                    .misbehave(session.id, misbehavior)
+                                    .is_disconnect()
+                                {
+                                    if context.disconnect(session.id).is_err() {
+                                        debug!("disconnect {:?} send fail", session.id)
+                                    }
+                                    return;
+                                }
+                            }
+
+                            let addrs = nodes
+                                .items
+                                .into_iter()
+                                .flat_map(|node| node.addresses.into_iter())
+                                .collect::<Vec<_>>();
+
+                            state.addr_known.extend(addrs.iter());
+                            state.received_nodes = true;
+
+                            self.addr_mgr.add_new_addrs(session.id, addrs);
+                        }
+                    }
+                }
+            }
+            None => {
+                if self
+                    .addr_mgr
+                    .misbehave(session.id, Misbehavior::InvalidData)
+                    .is_disconnect()
+                    && context.disconnect(session.id).is_err()
+                {
+                    debug!("disconnect {:?} send fail", session.id)
+                }
+            }
+        }
+    }
+
+    fn notify(&mut self, context: &mut ProtocolContext, _token: u64) {
+        let now = Instant::now();
+
+        let dynamic_query_cycle = self.dynamic_query_cycle.unwrap_or(ANNOUNCE_INTERVAL);
+        let addr_mgr = &self.addr_mgr;
+
+        // get announce list
+        let announce_list: Vec<_> = self
+            .sessions
+            .iter_mut()
+            .filter_map(|(id, state)| {
+                // send all announce addr to remote
```

Comment:
The function names are confusing here.

The method `send_messages` send the Nodes message as announce. And then we check the timer to update `state.announce`. It seems there are two different concepts of `announce`s. We can rename the latter `announce` to `state.to_be_anounced` to make it clear.